zcash_primitives/src/transaction/builder.rs

```diff
@@ -439,23 +452,25 @@ impl<'a, P: consensus::Parameters, R: RngCore + CryptoRng> Builder<'a, P, R> {
     ///
     /// Upon success, returns a tuple containing the final transaction, and the
     /// [`SaplingMetadata`] generated during the build process.
-    pub fn build<FR: FeeRule>(
+    pub fn build<SP: SpendProver, OP: OutputProver, FR: FeeRule>(
```

Comment:
> I feel like these should be qualified, and the corresponding variables named, to reflect their being specific to Sapling, given that this API is part of the top-level builder.
> [...]
> Is this PR just an intermediate step?

The type names _are_ qualified, with the `sapling::prover` module prefix. And when these eventually move into the `sapling-crypto` crate, we wouldn't want to include redundant prefixes on the type or variable names. At the top-level transaction builder however (that is staying in `zcash_primitives`) I'd be fine with a prefix.

Also note that the presence of these traits is an artifact of how the Sapling transaction builder was originally constructed to allow remote proving (in environments where local proving could be too costly), as well as different types to represent loading the parameters in different ways. Orchard doesn't have these intermediate traits; it instead has concrete types for its prover. But keeping with the trait-based approach kept this (and the prior PRs) simpler to implement and review.

So in this regard I do consider this PR to be an intermediate step: the `TxBuilder` trait is unused after this PR and can be removed, and similarly we could remove the ability to do remote proving (which AFAIK has never been done beyond a demo I put together years ago) and simplify the types to be more concrete like Orchard.

> Also, if the objective is to align the Sapling and Orchard designs, why does this need the Sapling prover(s) passed in but not an Orchard prover?

For the current transaction builder implementation, the Orchard proving parameters are being constructed on-the-fly. That adds overhead to the transaction building process that we won't want in wallets, but we haven't noticed it yet because we don't have Orchard support in `zcash_client_backend`. So yes, eventually we will want to pass in the Orchard proving parameters and let the caller decide when to build them.